src/descriptor/mod.rs

```diff
@@ -416,6 +418,31 @@ impl<P: ToPublicKey> Descriptor<P> {
     }
 }
 
+impl<P: ToPublicKey + Clone> Descriptor<P> {
+    pub fn interpret(
+        &self,
+        witness: &mut Vec<Vec<u8>>,
+        age: u32,
+        msg: &secp256k1::Message,
+        secp: &secp256k1::Secp256k1<All>,
+    ) -> Result<Primitives, Error>
+    {
+        match *self {
+            Descriptor::Bare(ref d) |
+            Descriptor::Sh(ref d) |
+            Descriptor::ShWsh(ref d) |
+            Descriptor::Wsh(ref d) => {
+                d.interpret(witness, age, msg, secp)
+            },
+            Descriptor::Pkh(ref pk) |
+            Descriptor::ShWpkh(ref pk) |
+            Descriptor::Wpkh(ref pk) => {
+                Ok(Primitives { pks: vec![pk.to_public_key()], hashlocks: vec![], timelocks: vec![] })
```

Comment:
Should the interpreter only check the signatures/hashes inside Miniscript or should it also interpret the entire descriptor and do a check for Pkh(P)? Right now, I am assuming that the descriptor has valid witnesses and only checking the interpretation of Miniscript.

Reply:
I think it should only check inside the miniscript.

I'd like to have another function which, given a scriptpubkey and witness script, figures out what kind of descriptor it is, but I need to think more about what that might take as input. But that function would be responsible for doing hashchecks.

Reply:
Yeah, so, it would be cool to have a function
```
pub fn witness_stack(
    script_pubkey: &bitcoin::Script,
    script_sig: &bitcoin::Script,
    witness: Vec<Vec<u8>>,
) -> Result<(Descriptor, Vec<StackElement>), Error>
```
where `StackElement` is an abstract stack element (I suggest a definition for this type in a later comment). This function would (a) check that the witness script is consistent with the scriptpubkey and error out otherwise; (b) figure out what kind of descriptor we're dealing with; (c) translate the `scriptSig` into a set of witness elements for `Bare`/`Pk`/`Pkh`/`Sh` descriptors; (d) drop the witness script from the stack for `Wsh`/`ShWsh` descriptors.

Reply:
I am slightly confused by the order in which things are written. I would need to determine (b) in order to solve (a), right?

Reply:
Yeah, sorry, (a) and (b) should be swapped.

Reply:
I have also added the signature validation logic alongside the hash checks for PublicKey based types, `Pkh`/`ShWpkh`/`ShWsh`.